NAAM.md

```diff
@@ -0,0 +1,278 @@
+
+* `ipns` multicodec code `0xe5` as protocol ID,
+* followed by marshalled IPNS Record.
+
+The use of metadata to capture IPNS records enables Naam to also utilise the built-in IPNI mechanism
+for updating metadata to update IPNS records. Because, context ID nor the multihash in Naam
+advertisements ever change for the same IPNS key. For more information,
+see [Updating IPNS Records](#updating-ipns-records).
+
+Note that the IPNS multicodec code `0xe5` is tagged as "namespace" in the CSV table, which could
+signify an IPNS path, not record. There is currently no multicodec code for IPNS records themselves.
+Alternatively, a new multicodec code can be introduced to the
+[multicodec CSV file](https://github.com/multiformats/multicodec/blob/master/table.csv) to avoid
+potential confusion between the two. The new code can then be used as the Protocol ID of Naam
+Metadata.
+
+### Updating IPNS Records
+
+IPNI offers a built-in mechanism to efficiently update the metadata associated to a list of
+multihashes without having to republish the entries. Indexer nodes uniquely identify the metadata by
+the advertisement `Provider`, and `ContextID`.
+
+Both of these values remain the same in Naam advertisements: the former is the same peer ID that
+forms part of the IPNS key, i.e. `/ipns/<peer-id>`, and the latter is fixed. As a result, to update
+an IPNS record, all that's needed is to publish a new Naam advertisement that:
+
+* links to the previous advertisement in `PrevousID` link as specified by IPNI,
+* uses the same `Provider` and `ContextID` values as before,
+* sets `IsRm` to `false`, and
+* has the updated IPNS record in Naam Metadata format.
+
+Note that there is no need to specify `Entries` when updating an existing record. It is inferred by
+the network indexers since `Provider` and `ContextID` in Naam advertisements never change.
+
+### Removing IPNS Records
+
+IPNS has EOL and TTL features built-in. This means explicit removal of records is not strictly
+required. Instead, records eventually expire. Though not strictly necessary, Naam protocol can
+utilise IPNI to explicitly remove records just like a traditional DNS nameserver.
+
+The process to remove an IPNS record published via Naam protocol is identical to removal
+advertisement specified by IPNI. Similar to a regular advertisement, an IPNS record may be removed
+by publishing an advertisement that:
+
+* links to the previous advertisement in `PrevousID` link as specified by IPNI,
+* uses the same `Provider` and `ContextID` values as before,
+* sets `IsRm` to `true`, and
+* sets `Entries` link to `NoEntries` as specified by IPNI.
+
+Once ingested, indexer nodes will eventually remove the Naam advertisement metadata along with the
+publisher information.
+
+### Resolving IPNS Records
+
+IPNS record resolution involves finding the IPNS record associated to a given IPNS key, i.e.
+`/ipns/<peer-id>`. As eluded to earlier, Naam uses the find API already provided by the network
```

Comment:
Was it meant to be `/ipns/<ipns-key>`?

Reply:
I thought `/ipns/<peer-id>` is more representative of what the keys [actually look like](https://github.com/ipfs/go-ipns/blob/master/record.go#L23).

Reply:
It is more correctly `<ipns-key>` which happens to be a peer ID. Something to note is that IMO one of the more powerful tools we have at the moment with the flexibility in the data storable with IPNI is the ability to make changes that would've taken a painfully long time otherwise (getting a DHT swarm to update takes much longer than getting a few IPNI nodes to update, or better yet just having the clients querying IPNI update).

This means that I would expect it to be much easier to evolve things like IPNS records.